modules/lang-painless/src/main/java/org/elasticsearch/painless/node/ECapturingFunctionRef.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.node;
+
+import org.elasticsearch.painless.DefBootstrap;
+import org.elasticsearch.painless.Definition;
+import org.elasticsearch.painless.FunctionRef;
+import org.elasticsearch.painless.Location;
+import org.elasticsearch.painless.MethodWriter;
+import org.elasticsearch.painless.Variables;
+import org.elasticsearch.painless.Variables.Variable;
+import org.objectweb.asm.Type;
+
+import static org.elasticsearch.painless.WriterConstants.DEF_BOOTSTRAP_HANDLE;
+import static org.elasticsearch.painless.WriterConstants.LAMBDA_BOOTSTRAP_HANDLE;
+
+import java.lang.invoke.LambdaMetafactory;
+
+/**
+ * Represents a capturing function reference.
+ */
+public class ECapturingFunctionRef extends AExpression {
+    public final String type;
+    public final String call;
+    
+    private FunctionRef ref;
+    Variable captured;
+    private boolean defInterface;
+
+    public ECapturingFunctionRef(Location location, String type, String call) {
+        super(location);
+
+        this.type = type;
+        this.call = call;
+    }
+
+    @Override
+    void analyze(Variables variables) {
+        captured = variables.getVariable(location, type);
+        if (expected == null) {
+            defInterface = true;
+            actual = Definition.getType("String");
+        } else {
+            defInterface = false;
+            // static case
+            if (captured.type.sort != Definition.Sort.DEF) {
+                try {
+                    ref = new FunctionRef(expected, captured.type.name, call, captured.type.clazz);
+                } catch (IllegalArgumentException e) {
+                    throw createError(e);
+                }
+            }
+            actual = expected;
+        }
+    }
+
+    @Override
+    void write(MethodWriter writer) {
+        writer.writeDebugInfo(location);
+        if (defInterface && captured.type.sort == Definition.Sort.DEF) {
+            // dynamic interface, dynamic implementation
+            writer.push("D" + type + "." + call + ",1");
+            writer.loadLocal(captured.slot);
```

Comment:
I don't think you can use this method because it won't store the type correctly since we do the slots ourself to avoid trash being on the stack with variables scopes and such.  Instead you'll have to use writer.visitVarInsn(asmtype.getOpcode(Opcodes.ILOAD), slot);